pkg/scheduler/plugins/loadaware/pod_assign_cache.go

```diff
@@ -211,3 +230,264 @@ func (p *podAssignCache) OnDelete(obj interface{}) {
+	}
+	info := metric.Status.NodeMetric
+	if info != nil {
+		m.nodeUsage = p.vectorizer.ToVec(info.NodeUsage.ResourceList)
+		if aggLen := len(info.AggregatedNodeUsages); aggLen > 0 {
+			typeLen := len(info.AggregatedNodeUsages[0].Usage)
+			aggUsages := make(map[aggUsageKey]ResourceVector, (aggLen+1)*typeLen)
+			maxDurations := make(map[extension.AggregationType]time.Duration, typeLen)
+			for _, aggInfo := range info.AggregatedNodeUsages {
+				d := aggInfo.Duration.Duration
+				for t, u := range aggInfo.Usage {
+					if len(u.ResourceList) == 0 {
+						continue
+					}
+					key := aggUsageKey{Type: t, Duration: d}
+					vec := p.vectorizer.ToVec(u.ResourceList)
+					aggUsages[key] = vec
+					if md := maxDurations[t]; d > md {
+						maxDurations[t] = d
+					}
+				}
+			}
+			for t, d := range maxDurations {
+				aggUsages[aggUsageKey{Type: t}] = aggUsages[aggUsageKey{Type: t, Duration: d}]
+			}
+			m.aggUsages = aggUsages
+		}
+		if p.args.ProdUsageIncludeSys {
+			m.prodUsage.Add(p.vectorizer.ToVec(info.SystemUsage.ResourceList))
+		}
+	}
+	if infos := metric.Status.PodsMetric; len(infos) != 0 {
+		podUsages := make(map[NamespacedName]ResourceVector, len(infos))
+		prodPods := sets.New[NamespacedName]()
+		for _, info := range infos {
+			if info == nil {
+				continue
+			}
+			if len(info.PodUsage.ResourceList) == 0 {
+				continue
+			}
+			key := NamespacedName{Namespace: info.Namespace, Name: info.Name}
+			vec := p.vectorizer.ToVec(info.PodUsage.ResourceList)
+			podUsages[key] = vec
+			if info.Priority == extension.PriorityProd {
+				prodPods.Insert(key)
+			}
+		}
+		m.podUsages, m.prodPods = podUsages, prodPods
+	}
+	return m
+}
+
+func (p *podAssignCache) initPods(m *nodeMetric, pods map[types.UID]*podAssignInfo) {
+	m.nodeDelta = p.vectorizer.EmptyVec()
+	m.prodDelta = p.vectorizer.EmptyVec()
+	m.nodeEstimated = p.vectorizer.EmptyVec()
+
+	if klog.V(6).Enabled() {
+		m.nodeDeltaPods = sets.New[NamespacedName]()
+		m.prodDeltaPods = sets.New[NamespacedName]()
+		m.nodeEstimatedPods = sets.New[NamespacedName]()
+	}
+
+	for _, pod := range pods {
+		p.addPod(m, pod)
+	}
+}
+
+func (p *podAssignCache) addPod(m *nodeMetric, pod *podAssignInfo) {
+	key := NamespacedName{Namespace: pod.pod.Namespace, Name: pod.pod.Name}
+	u := m.podUsages[key]
+	prod := extension.GetPodPriorityClassWithDefault(pod.pod) == extension.PriorityProd
+	// Only use prod pod's usage when both pod claims and node metric reports it as prod.
+	// 1. pod priority class are updated dynamically
+	// 2. prod / non prod pod is wrongly reported or terminated pod is leaked in node metrics status
+	activeProd := prod && m.prodPods.Has(key)
+	if activeProd {
+		m.prodUsage.Add(u)
+	}
+
+	e := pod.estimated
+	if e == nil {
+		return
+	}
+	// 1. when usage is not collected
+	// 2. when pod miss lastest metrics update
+	// 3. when pod metrics is still in the report interval
+	// 4. when pod is configured in estimation
+	should := u == nil ||
```

Comment:
The load-aware plugin requires real-time metrics during the scheduling cycle. But this judgment only triggers calculation upon new events, which breaks the timeliness of the estimation.

Reply:
Actually, in the real time calculation code only estimatedDeadline check is depend on current timestamp, which is newly added and okay to replace with metrics updateTime.

Other data are all from cache: existing pods are fetched from this pod assign cache, node metrics are fetched from informer cache.